File: language/src/main/java/cadenza/types/Term.java
package cadenza.types;

import cadenza.nodes.Code;
import com.oracle.truffle.api.frame.FrameDescriptor;

import java.util.Arrays;

import static cadenza.nodes.Code.*;

// terms can be checked and inferred. The result is an expression.
public abstract class Term {
  // expected is optional for some types, giving us bidirectional type checking.
  public abstract Witness check(Ctx ctx, Type expected) throws TypeError;
  public final Witness infer(Ctx ctx) throws TypeError { return check(ctx, null); }

    public static Term tname(String name) {
    return new Term() {
      @Override
      public Witness check(Ctx ctx, Type _expectedType) throws TypeError {
        return new Witness(lookup(ctx,name)) {
          @Override public Code compile(FrameDescriptor fd) {
            return Code.var(fd.findOrAddFrameSlot(name));
          }
        };
      }
    };
  }

  public static Term tif(Term body, Term thenTerm, Term elseTerm) {
    return new Term() {
      @Override
      public Witness check(Ctx ctx, Type expectedType) throws TypeError {
        var bodyWitness = body.check(ctx, Type.bool);
        var thenWitness = thenTerm.check(ctx, expectedType);
        var actualType = thenWitness.type;
        var elseWitness = elseTerm.check(ctx, actualType);
        return new Witness(actualType) {
          @Override public Code compile(FrameDescriptor fd) {
            return new Code.If(actualType, bodyWitness.compile(fd), thenWitness.compile(fd), elseWitness.compile(fd));
          }
        };
      }
    };
  }

  public static Term tapp(Term trator, Term... trands) {
    return new Term() {
      @Override
      public Witness check(Ctx ctx, Type expectedType) throws TypeError {
        var wrator = trator.check(ctx, expectedType);
        var currentType = wrator.type;
        int len = trands.length;
        var wrands = new Witness[len]; // can't make into a nice stream, throws type errors
        for (int i = 0; i < len; ++i) {
          var arr = (Type.Arr) currentType;
          if (arr == null) throw new TypeError("not a fun type");
          wrands[i] = trands[i].check(ctx, arr.argument);
          currentType = arr.result;
        }
        return new Witness(currentType) {
          @Override public Code compile(FrameDescriptor fd) {
            return app(wrator.compile(fd), Arrays.stream(wrands).map(w -> w.compile(fd)).toArray(Code[]::new));
          }
        };
      }
    };
  }

  public static Term tlam(String[] _names, Term _body) {
    return null;
  }

  // provides an expression with a given type in a given frame
  public static abstract class Witness {
    public final Type type;
    Witness(Type type) { this.type = type; }
    public abstract Code compile(FrameDescriptor fd); // take a frame descriptor and emit an expression

    // builder style
    public Witness match(Type expectedType) throws TypeError {
      if (expectedType != null && type != expectedType)
        throw new TypeError("type mismatch", type, expectedType);
      return this;
    }
  }

  // singly linked list
  public static final class Ctx {
    public final String name;
    public final Type type;
    public final Ctx next;

    public Ctx(final String name, final Type type, final Ctx next) {
      this.name = name;
      this.type = type;
      this.next = next;
    }

    public static Ctx nil = null;
    public static Ctx cons(String name, Type type, Ctx next) {
      return new Ctx(name,type,next);
    }
  }

  static Type lookup(final Ctx ctx, final String name) throws TypeError {
    for(Ctx current = ctx; current != null; current = current.next) {
      if (name.equals(current.name)) return current.type;
    }
    throw new TypeError("unknown variable");
  }

}

File: language/src/main/kotlin/cadenza/types/Term.kt
package cadenza.types

import cadenza.nodes.Code
import com.oracle.truffle.api.frame.FrameDescriptor

import java.util.Arrays

import cadenza.nodes.Code.*

// terms can be checked and inferred. The result is an expression.
abstract class Term {
  // expected is optional for some types, giving us bidirectional type checking.
  @Throws(TypeError::class)
  abstract fun check(ctx: Ctx, expected: Type?): Witness

  @Throws(TypeError::class)
  fun infer(ctx: Ctx): Witness {
    return check(ctx, null)
  }

  // provides an expression with a given type in a given frame
  abstract class Witness internal constructor(val type: Type) {
    abstract fun compile(fd: FrameDescriptor): Code  // take a frame descriptor and emit an expression

    // builder style
    @Throws(TypeError::class)
    fun match(expectedType: Type?): Witness {
      if (expectedType != null && type !== expectedType)
        throw TypeError("type mismatch", type, expectedType)
      return this
    }
  }

  // singly linked list
  class Ctx(val name: String, val type: Type, val next: Ctx) {
    companion object {

      var nil: Ctx? = null
      fun cons(name: String, type: Type, next: Ctx): Ctx {
        return Ctx(name, type, next)
      }
    }
  }

  companion object {

    fun tname(name: String): Term {
      return object : Term() {
        @Throws(TypeError::class)
        override fun check(ctx: Ctx, _expectedType: Type?): Witness {
          return object : Witness(lookup(ctx, name)) {
            override fun compile(fd: FrameDescriptor): Code {
              return Code.`var`(fd.findOrAddFrameSlot(name))
            }
          }
        }
      }
    }

    fun tif(body: Term, thenTerm: Term, elseTerm: Term): Term {
      return object : Term() {
        @Throws(TypeError::class)
        override fun check(ctx: Ctx, expectedType: Type?): Witness {
          val bodyWitness = body.check(ctx, Type.Bool)
          val thenWitness = thenTerm.check(ctx, expectedType)
          val actualType = thenWitness.type
          val elseWitness = elseTerm.check(ctx, actualType)
          return object : Witness(actualType) {
            override fun compile(fd: FrameDescriptor): Code {
              return Code.If(actualType, bodyWitness.compile(fd), thenWitness.compile(fd), elseWitness.compile(fd))
            }
          }
        }
      }
    }

    fun tapp(trator: Term, vararg trands: Term): Term {
      return object : Term() {
        @Throws(TypeError::class)
        override fun check(ctx: Ctx, expectedType: Type?): Witness {
          val wrator = trator.check(ctx, expectedType)
          var currentType = wrator.type
          val len = trands.size
          val wrands = trands.map {
            val arr = currentType as Type.Arr ?: throw TypeError("not a fun type")
            val out = it.check(ctx, arr.argument)
            currentType = arr.result
            return out
          }.toTypedArray<Witness>()
          return object : Witness(currentType) {
            override fun compile(fd: FrameDescriptor): Code {
              return App(
                wrator.compile(fd),
                wrands.map { it.compile(fd) }.toTypedArray()
              )
            }
          }
        }
      }
    }

    fun tlam(_names: Array<String>, _body: Term): Term? {
      return null
    }

    @Throws(TypeError::class)
    internal fun lookup(ctx: Ctx, name: String): Type {
      var current: Ctx? = ctx
      while (current != null) {
        if (name == current.name) return current.type
        current = current.next
      }
      throw TypeError("unknown variable")
    }
  }

}
